Packaging Springboot 3 application embedded with Open Liberty does not work

**Diagnostic information:**  
 - OpenLiberty Version: 23.0.0.9
 - Affected feature(s): springBoot-3.0
 - Java Version: 17

**Problem**
- Configure the pom for minify and runnable
```
      <plugin>
        <groupId>io.openliberty.tools</groupId>
        <artifactId>liberty-maven-plugin</artifactId>
        <version>3.8.2</version>
        <configuration>
          <appsDirectory>apps</appsDirectory>
          <installAppPackages>spring-boot-project</installAppPackages>
          <include>minify,runnable</include>
          <packageName>GSSpringBootApp</packageName>
        </configuration>
        <executions>
          <execution>
            <id>package-server</id>
            <phase>package</phase>
            <goals>
              <goal>create</goal>
              <goal>install-feature</goal>
              <goal>deploy</goal>
              <goal>package</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
```
- server.xml
```
    <featureManager>
        <feature>servlet-6.0</feature>
        <feature>springBoot-3.0</feature>
    </featureManager>

    <httpEndpoint id="defaultHttpEndpoint"
                  host="*"
                  httpPort="9080"
                  httpsPort="9443" />

    <springBootApplication id="guide-spring-boot" 
                           location="thin-guide-spring-boot-0.1.0.jar"
                           name="guide-spring-boot" />
```
- `./mvnw package`
  - it generates target/GSSpringBootApp.jar
- `java -jar target/GSSpringBootApp.jar`
  - got following error
```
gkwan@gilberts-mbp-2 start % java -jar target/GSSpringBootApp.jar
Extracting files to /Users/gkwan/wlpExtract/GSSpringBootApp_68559353251333/wlp
Successfully extracted all product files.
Launching defaultServer (Open Liberty 23.0.0.9/wlp-1.0.81.cl230920230904-1158) on Eclipse OpenJ9 VM, version 17.0.6+10 (en_CA)
[AUDIT   ] CWWKE0001I: The server defaultServer has been launched.
[AUDIT   ] CWWKG0093A: Processing configuration drop-ins resource: /Users/gkwan/wlpExtract/GSSpringBootApp_68559353251333/wlp/usr/servers/defaultServer/configDropins/defaults/install_apps_configuration_1491924271.xml
[AUDIT   ] CWWKZ0058I: Monitoring dropins for applications.
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v3.1.0)
2023-09-28T15:22:21.311-04:00  INFO 90897 --- [ecutor-thread-6] com.example.springboot.Application       : Starting Application using Java 17.0.6 with PID 90897 (/Users/gkwan/wlpExtract/GSSpringBootApp_68559353251333/wlp/usr/servers/defaultServer/apps/thin-guide-spring-boot-0.1.0.jar started by gkwan in /Users/gkwan/wlpExtract/GSSpringBootApp_68559353251333/wlp/usr/servers/defaultServer)
2023-09-28T15:22:21.316-04:00  INFO 90897 --- [ecutor-thread-6] com.example.springboot.Application       : No active profile set, falling back to 1 default profile: "default"
2023-09-28T15:22:22.236-04:00  WARN 90897 --- [ecutor-thread-6] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.context.ApplicationContextException: Unable to start web server
2023-09-28T15:22:22.260-04:00  INFO 90897 --- [ecutor-thread-6] .s.b.a.l.ConditionEvaluationReportLogger : 

Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
2023-09-28T15:22:22.275-04:00 ERROR 90897 --- [ecutor-thread-6] o.s.boot.SpringApplication               : Application run failed

org.springframework.context.ApplicationContextException: Unable to start web server
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:164) ~[spring-boot-3.1.0.jar:3.1.0]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:602) ~[spring-context-6.0.9.jar:6.0.9]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.1.0.jar:3.1.0]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:733) ~[spring-boot-3.1.0.jar:3.1.0]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:435) ~[spring-boot-3.1.0.jar:3.1.0]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:311) ~[spring-boot-3.1.0.jar:3.1.0]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1305) ~[spring-boot-3.1.0.jar:3.1.0]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1294) ~[spring-boot-3.1.0.jar:3.1.0]
	at com.example.springboot.Application.main(Application.java:15) ~[thin-guide-spring-boot-0.1.0.jar:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at com.ibm.ws.app.manager.springboot.internal.SpringBootRuntimeContainer.lambda$invokeSpringMain$6(SpringBootRuntimeContainer.java:139) ~[na:na]
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:247) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:857) ~[na:na]
Caused by: java.lang.NoClassDefFoundError: io.openliberty.checkpoint.spi.CheckpointPhase
	at io.openliberty.springboot.support.web.server.version30.container.LibertyServletWebServerFactory.getWebServer(LibertyServletWebServerFactory.java:36) ~[springBoot30Server.jar:na]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:183) ~[spring-boot-3.1.0.jar:3.1.0]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:161) ~[spring-boot-3.1.0.jar:3.1.0]
	... 17 common frames omitted
Caused by: java.lang.ClassNotFoundException: io.openliberty.checkpoint.spi.CheckpointPhase
	at com.ibm.ws.classloading.internal.AppClassLoader.findClassCommonLibraryClassLoaders(AppClassLoader.java:736) ~[na:na]
	at [internal classes]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1094) ~[na:na]
	... 20 common frames omitted
[ERROR   ] CWWKZ0002E: An exception occurred while starting the application guide-spring-boot. The exception message was: org.springframework.context.ApplicationContextException: Unable to start web server
[AUDIT   ] CWWKF0012I: The server installed the following features: [servlet-6.0, springBoot-3.0].
[AUDIT   ] CWWKF0011I: The defaultServer server is ready to run a smarter planet. The defaultServer server started in 3.920 seconds.
```

--- a/dev/com.ibm.ws.app.manager.springboot/src/com/ibm/ws/app/manager/springboot/container/SpringBootConfig.java
+++ b/dev/com.ibm.ws.app.manager.springboot/src/com/ibm/ws/app/manager/springboot/container/SpringBootConfig.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -15,6 +15,8 @@ package com.ibm.ws.app.manager.springboot.container;
 import com.ibm.ws.app.manager.springboot.container.config.ServerConfiguration;
 import com.ibm.ws.app.manager.springboot.container.config.SpringConfiguration;
 import com.ibm.ws.app.manager.springboot.support.ContainerInstanceFactory;
+
+import io.openliberty.checkpoint.spi.CheckpointPhase;
 
 /**
  * Spring Boot configuration that configures a embedded container such as a
@@ -60,4 +62,9 @@ public interface SpringBootConfig {
      * @return Spring boot config id
      */
     String getId();
+
+    public static boolean isBeforeCheckpoint() {
+        CheckpointPhase phase = CheckpointPhase.getPhase();
+        return !phase.restored();
+    }
 }

--- a/dev/io.openliberty.springboot.support.web.server.version30/src/io/openliberty/springboot/support/web/server/version30/container/LibertyReactiveWebServerFactory.java
+++ b/dev/io.openliberty.springboot.support.web.server.version30/src/io/openliberty/springboot/support/web/server/version30/container/LibertyReactiveWebServerFactory.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -17,12 +17,14 @@ import org.springframework.boot.SpringBootVersion;
 import org.springframework.boot.context.properties.ConfigurationProperties;
 import org.springframework.boot.web.reactive.server.AbstractReactiveWebServerFactory;
 import org.springframework.boot.web.server.WebServer;
+import org.springframework.boot.web.server.WebServerException;
 import org.springframework.boot.web.servlet.ServletContextInitializer;
 import org.springframework.context.ApplicationContext;
 import org.springframework.context.ApplicationContextAware;
 import org.springframework.http.server.reactive.HttpHandler;
 import org.springframework.http.server.reactive.ServletHttpHandlerAdapter;
 
+import com.ibm.ws.app.manager.springboot.container.SpringBootConfig;
 import com.ibm.ws.springboot.support.web.server.initializer.ServerConfigurationFactory;
 
 import jakarta.servlet.ServletRegistration;
@@ -46,7 +48,36 @@ public class LibertyReactiveWebServerFactory extends AbstractReactiveWebServerFa
             registration.addMapping("/");
             registration.setAsyncSupported(true);
         } };
-        return new LibertyWebServer(this, this, initializers);
+        if (SpringBootConfig.isBeforeCheckpoint()) {
+            // for InstantOn we create a wrapper so that we can recreate the LibertyWebServer on restart
+            return new WebServer() {
+                LibertyWebServer webServer = new LibertyWebServer(LibertyReactiveWebServerFactory.this, LibertyReactiveWebServerFactory.this, initializers);
+
+                @Override
+                public synchronized void start() throws WebServerException {
+                    if (webServer == null) {
+                        webServer = new LibertyWebServer(LibertyReactiveWebServerFactory.this, LibertyReactiveWebServerFactory.this, initializers);
+                    }
+                    webServer.start();
+                }
+
+                @Override
+                public synchronized void stop() throws WebServerException {
+                    if (webServer != null) {
+                        webServer.stop();
+                        webServer = null;
+                    }
+                }
+
+                @Override
+                public synchronized int getPort() {
+                    return webServer != null ? webServer.getPort() : 0;
+                }
+            };
+        } else {
+            return new LibertyWebServer(this, this, initializers);
+        }
+
     }
 
     @Override

--- a/dev/io.openliberty.springboot.support.web.server.version30/src/io/openliberty/springboot/support/web/server/version30/container/LibertyServletWebServerFactory.java
+++ b/dev/io.openliberty.springboot.support.web.server.version30/src/io/openliberty/springboot/support/web/server/version30/container/LibertyServletWebServerFactory.java
@@ -21,7 +21,7 @@ import org.springframework.boot.web.servlet.server.AbstractServletWebServerFacto
 import org.springframework.context.ApplicationContext;
 import org.springframework.context.ApplicationContextAware;
 
-import io.openliberty.checkpoint.spi.CheckpointPhase;
+import com.ibm.ws.app.manager.springboot.container.SpringBootConfig;
 
 /**
  *
@@ -33,8 +33,7 @@ public class LibertyServletWebServerFactory extends AbstractServletWebServerFact
 
     @Override
     public WebServer getWebServer(ServletContextInitializer... initializers) {
-        CheckpointPhase phase = CheckpointPhase.getPhase();
-        if (!phase.restored()) {
+        if (SpringBootConfig.isBeforeCheckpoint()) {
             // for InstantOn we create a wrapper so that we can recreate the LibertyWebServer on restart
             return new WebServer() {
                 LibertyWebServer webServer = new LibertyWebServer(LibertyServletWebServerFactory.this, LibertyServletWebServerFactory.this, mergeInitializers(initializers));
@@ -42,7 +41,7 @@ public class LibertyServletWebServerFactory extends AbstractServletWebServerFact
                 @Override
                 public synchronized void start() throws WebServerException {
                     if (webServer == null) {
-                        webServer = new LibertyWebServer(LibertyServletWebServerFactory.this, LibertyServletWebServerFactory.this, mergeInitializers(initializers));;
+                        webServer = new LibertyWebServer(LibertyServletWebServerFactory.this, LibertyServletWebServerFactory.this, mergeInitializers(initializers));
                     }
                     webServer.start();
                 }

